### PR TITLE
Add emission class for Categorical distributions

### PR DESCRIPTION
**Pull request by ThomasColthurst**
A later pull request will have the upgrades to the Sometimes emissions adapter that will let it be used in conjunction with this.

### CODE REVIEW

File: cxx/emissions/categorical.hh [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    emission_dists[x.first].unincorporate(x.second);
+  }
+
+  double logp(const std::pair<int, int>& x) const {
+    double lp;
+    for (size_t i = 0; i < emission_dists.size(); ++i) {
+      if (std::cmp_equal(i, x.first)) {
+        lp += emission_dists[i].logp(x.second);
+      } else {
+        lp += emission_dists[i].logp_score();
```
Comment (emilyfertig) [inline cxx/emissions/categorical.hh:39]: Could you explain this formula for logp? I thought it should be `p(x.second | x.first)`, where `x.first` is clean and `x.second` is noisy, such that the conditional probability in the `if` branch was all we needed. Also, to the extent that this is like a bigram model, the bigram `logp` doesn't contain `logp_score` terms for the rest of the categories.

**Review by emilyfertig — COMMENTED** [review]
I have a question about logp, otherwise looks good!

File: cxx/emissions/categorical.hh (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/emissions/categorical.hh]: You are entirely correct; this should only contain the conditional probability.  Fixed.

**Review by emilyfertig — APPROVED** [review]
Thanks, LG!

**Pull request merged**